src/lib/Embed.hs

```diff
@@ -188,6 +190,28 @@ buildNAbsAux bs body = do
      return (fmap Bind vs, result)
   return (Abs bs' $ wrapDecls decls ans, aux)
 
+buildDataDef :: MonadEmbed m
+             => Name -> Nest Binder -> ([Atom] -> m [DataConDef]) -> m DataDef
+buildDataDef tyConName paramBinders body = do
+  ((paramBinders', dataDefs), _) <- scopedDecls $ do
+     vs <- freshNestedBinders paramBinders
+     result <- body $ map Var $ toList vs
+     return (fmap Bind vs, result)
+  return $ DataDef tyConName paramBinders' dataDefs
+
+buildImplicitNaryLam :: MonadEmbed m => (Nest Binder) -> ([Atom] -> m Atom) -> m Atom
+buildImplicitNaryLam Empty body = body []
+buildImplicitNaryLam (Nest b bs) body =
+  buildLam b ImplicitArrow $ \x -> do
+    bs' <- substEmbed (b@>x) bs
```

Comment:
nit: this is O(n^2). Would be better to carry the substitution in a recursive call, but no big deal I guess

Reply:
Good catch. I think there are many places like this in the compiler. One day we'll need to do some profiling and fix the ones that matter the most. I'll leave this one for now.